### PR TITLE
Add SM2 Algorithm Support

### PR DESCRIPTION
**Pull request by xicilion**
## Overview
Add SM2 (ShangMi 2) cryptographic algorithm support to the controller document specification. SM2 is a public key cryptographic algorithm based on elliptic curves, which has been standardized as GM/T 0003-2012 by the State Cryptography Administration of China.

## Changes
- Added SM2 algorithm specification and parameters
- Included SM2 in the supported cryptographic algorithms list
- Added relevant documentation and examples

## Motivation
SM2 is widely used in China's commercial cryptographic applications and has been proven to provide strong security. Adding SM2 support enhances the specification's compatibility with Chinese cryptographic standards and provides more options for implementers requiring compliance with Chinese regulations.

## Impact
This addition expands the cryptographic algorithm options available to implementers while maintaining backward compatibility with existing implementations.

## Related Issues
N/A

***
<a href="https://pr-preview.s3.amazonaws.com/xicilion/controller-document/pull/131.html" title="Last updated on Dec 18, 2024, 4:59 PM UTC (2bd3269)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/cid/131/9b500df...xicilion:2bd3269.html" title="Last updated on Dec 18, 2024, 4:59 PM UTC (2bd3269)">Diff</a>

### CODE REVIEW

**Review by msporny — APPROVED** [review]

**Review by iherman — APPROVED** [review]
I am not familiar with the details of SM2 to judge whether the definition is mathematically correct. But I approve the principle to include this.

**Review by dlongley — APPROVED** [review]

**Review by PatStLouis — APPROVED** [review]

File: index.html [hunk 1]
```diff
@@ -3764,6 +3784,16 @@ <h2>Multikey Examples</h2>
   "capabilityInvocation": [
     "https://controller.example/123#key-2"
   ]
+}
+          </pre>
+
+          <pre class="example nohighlight"
+            title="A SM2 public key encoded as a Multikey">
```
Comment (TallTed) [inline index.html:3791]:
```suggestion
            title="An SM2 public key encoded as a Multikey">
```

File: index.html [hunk 2]
```diff
@@ -3882,6 +3912,21 @@ <h2>JsonWebKey Examples</h2>
   "capabilityInvocation": [
     "https://controller.example/123#key-2"
   ]
+}
+          </pre>
+
+          <pre class="example nohighlight"
+            title="A SM2 public key encoded as a JsonWebKey">
```
Comment (TallTed) [inline index.html:3919]:
```suggestion
            title="An SM2 public key encoded as a JsonWebKey">
```

**Review by TallTed — CHANGES_REQUESTED** [review]
LGTM after these `a` become `an`

File: index.html [hunk 3]
```diff
@@ -1435,6 +1435,16 @@ <h3>Multikey</h3>
 (`z`).
                   </td>
                 </tr>
+                <tr>
+                  <td>SM2 256-bit public key</td>
+                  <td>
+The encoding of a SM2 public key MUST start with the two-byte prefix
```
Comment (TallTed) [inline index.html:1441]:
```suggestion
The encoding of an SM2 public key MUST start with the two-byte prefix
```

File: index.html [hunk 4]
```diff
@@ -1489,6 +1499,16 @@ <h3>Multikey</h3>
 (`z`).
                   </td>
                 </tr>
+                <tr>
+                  <td>SM2 256-bit secret key</td>
+                  <td>
+The encoding of a SM2 secret key MUST start with the two-byte prefix
```
Comment (TallTed) [inline index.html:1505]:
```suggestion
The encoding of an SM2 secret key MUST start with the two-byte prefix
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Sorry, missed two other locations of "a SM2" that should be "an SM2"

**Review by selfissued — CHANGES_REQUESTED** [review]
I believe there is not a registered algorithm identifier for SM2 in https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms so there is no defined JOSE representation for this algorithm.

**Comment by xicilion** [comment]
It is true that SM2 is not currently supported by JOSE and there is already a community proposal to suggest this feature:

https://www.ietf.org/archive/id/draft-dang-webauthn-sm2-00.html

But it doesn't seem to have been accepted. But I believe this should not be a barrier, the BLS12381G2 algorithm is also not on the JOSE list, which does not prevent did from using BLS12-381 as one of the optional signature algorithms.

As one of the technical barriers is the definition of MultiCodoc, I have previously initiated proposals to define multibase prefixes for public and private keys for SM2.

https://github.com/multiformats/multicodec/pulls?q=is%3Apr+is%3Aclosed+sm2

**Comment by xicilion** [comment]
> Sorry, missed two other locations of "a SM2" that should be "an SM2"

No, that was my mistake, I misunderstood you and thought you only wished to modify the two previous ones.

**Comment by selfissued** [comment]
I suggest submitting a new draft either called draft-dang-jose-sm2 or draft-dang-cose-sm2, depending upon which working group you want to consider it.  (There isn't an IETF WebAuthn working group.)  And then request presentation time at IETF 122 in Bangkok to ask for the draft to be considered by the working group.

**Comment by xicilion** [comment]
Thank you very much for your suggestions, I will spend some time figuring out how to submit this draft.

File: index.html [hunk 5]
```diff
@@ -3882,6 +3912,21 @@ <h2>JsonWebKey Examples</h2>
...
+{
+  "id": "https://jsonwebkey.example/issuer/123#key-0",
+  "type": "JsonWebKey",
+  "controller": "https://jsonwebkey.example/issuer/123",
+  "publicKeyJwk": {
+    "kty": "EC",
+    "crv": "SM2",
+    "x": "NIMpjy0M25izi2DrQLhQBlL8b45Lt9a0FzfELdpdO0s",
+    "y": "NXstrSPtSp1cZzvupZ8C68FCgSQhZbuGYLorP6F1N2w"
+  }
```
Comment (msporny) [inline index.html:3929]:
```suggestion
```

@selfissued noted that JsonWebKey does not officially support SM2 and has requested its removal so this PR can be merged.

**Review by TallTed — APPROVED** [review]
My tweaks were made. English looks good. fwiw

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2024-12-18-vcwg#section2-3) on 2024-12-18

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.3. Add SM2 Algorithm Support (pr cid#131)

_See github pull request [cid#131](https://github.com/w3c/cid/pull/131)._




**Brent Zundel:** add SM2 algorithm support. Has a couple requests for changes.  

**Manu Sporny:** I put a change suggestion in there to see if the original person that raised the issue would have any objections. If they don't, we can merge it after. I'll remove the JSON web key expression and merge it after a day or two of waiting.  

**Brent Zundel:** I believe the changes requested by TallTed have gone in. Path forward is the SM2 algorithm that's being added will be added for Data Integrity, on the JWK side it would be handled at IETF.  
… with that all concerns will have been addressed and the PR can be merged. If folks disagree with this plan please speak up, otherwise we will move to our final PR.  

</details>

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, merging.

**Comment by xicilion** [comment]
Great. Thank you all for your help.

**Comment by caict-develop-zhangbo** [comment]
没有一起把SM3 提交了

**Comment by iherman** [comment]
> 没有一起把SM3 提交了

The google translation says: "Didn't submit SM3 together". I.e., the question may be why SM3 has not been submitted. 

I am not familiar with all the details, but, afaik, SM3 is not the same as SM2, meaning that it would require a separate submission and appear in the final list of entries alongside SM2.

**Comment by caict-develop-zhangbo** [comment]
> > 没有一起把SM3 提交了
> 
> The google translation says: "Didn't submit SM3 together". I.e., the question may be why SM3 has not been submitted.
> 
> I am not familiar with all the details, but, afaik, SM3 is not the same as SM2, meaning that it would require a separate submission and appear in the final list of entries alongside SM2.

Thank you for your reply. SM3 is a hash algorithm of the Chinese standard. Many of our DID methods utilize SM3. I noticed that the Multihash property does not include this value, so I was asking the PR author if they had any intention of submitting it. Perhaps I can initiate the SM3 PR.

**Comment by iherman** [comment]
Practicalities: we are in the process of releasing a final recommendation very very soon. This means that we are, essentially, in feature freeze, and I am afraid any such PR will probably be labeled as "future".

cc @msporny

**Comment by msporny** [comment]
> Practicalities: we are in the process of releasing a final recommendation very very soon. This means that we are, essentially, in feature freeze, and I am afraid any such PR will probably be labeled as "future".

Yes, that is quite unfortunate that SM3 wasn't submitted earlier. We really should pull Multihash, Multibase, and Multikey into their own specifications and operate them as registries so we don't have this issue in the future.

@caict-develop-zhangbo while it is unlikely that we can add SM3 right now, we can add it in the v1.1 specification in a few months, or (ideally) we split the Multiformats specifications into their own specifications and establish a registry to avoid issues like this in the future.

**Pull request merged**